framework/crud_resource.go

```diff
@@ -0,0 +1,420 @@
+			return nil
+		}
+		if resourcecanceledcontext.IsCanceled(ctx) {
+			ctx = resourcecanceledcontext.NewContext(ctx, make(chan struct{}))
+			return nil
+		}
+
+		meta, ok := loggermeta.FromContext(ctx)
+		if ok {
+			meta.KeyVals["function"] = "GetDesiredState"
+			defer delete(meta.KeyVals, "function")
+		}
+		desiredState, err = r.ops.GetDesiredState(ctx, obj)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	var patch *Patch
+	{
+		if reconciliationcanceledcontext.IsCanceled(ctx) {
+			return nil
+		}
+		if resourcecanceledcontext.IsCanceled(ctx) {
+			ctx = resourcecanceledcontext.NewContext(ctx, make(chan struct{}))
+			return nil
+		}
+
+		meta, ok := loggermeta.FromContext(ctx)
+		if ok {
+			meta.KeyVals["function"] = "NewDeletePatch"
+			defer delete(meta.KeyVals, "function")
+		}
+		patch, err = r.ops.NewDeletePatch(ctx, obj, currentState, desiredState)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	}
+
+	{
+		if reconciliationcanceledcontext.IsCanceled(ctx) {
+			return nil
+		}
+		if resourcecanceledcontext.IsCanceled(ctx) {
+			ctx = resourcecanceledcontext.NewContext(ctx, make(chan struct{}))
+			return nil
+		}
+
+		if patch != nil {
+			createChange, ok := patch.getCreateChange()
+			if ok {
+				meta, ok := loggermeta.FromContext(ctx)
+				if ok {
+					meta.KeyVals["function"] = "ApplyCreateChange"
+					defer delete(meta.KeyVals, "function")
+				}
+				err := r.ops.ApplyCreateChange(ctx, obj, createChange)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			}
+		}
+	}
+
+	{
+		if reconciliationcanceledcontext.IsCanceled(ctx) {
+			return nil
+		}
+		if resourcecanceledcontext.IsCanceled(ctx) {
+			ctx = resourcecanceledcontext.NewContext(ctx, make(chan struct{}))
+			return nil
+		}
+
+		if patch != nil {
+			deleteChange, ok := patch.getDeleteChange()
+			if ok {
+				meta, ok := loggermeta.FromContext(ctx)
+				if ok {
+					meta.KeyVals["function"] = "ApplyDeleteChange"
+					defer delete(meta.KeyVals, "function")
+				}
+				err := r.ops.ApplyDeleteChange(ctx, obj, deleteChange)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			}
+		}
+	}
+
+	{
+		if reconciliationcanceledcontext.IsCanceled(ctx) {
+			return nil
+		}
+		if resourcecanceledcontext.IsCanceled(ctx) {
+			ctx = resourcecanceledcontext.NewContext(ctx, make(chan struct{}))
+			return nil
+		}
+
+		if patch != nil {
+			updateChange, ok := patch.getUpdateChange()
+			if ok {
+				meta, ok := loggermeta.FromContext(ctx)
+				if ok {
+					meta.KeyVals["function"] = "ApplyUpdateChange"
+					defer delete(meta.KeyVals, "function")
+				}
+				err := r.ops.ApplyUpdateChange(ctx, obj, updateChange)
+				if err != nil {
+					return microerror.Mask(err)
+				}
+			}
+		}
+	}
+
+	return nil
+}
+
+func (r *CRUDResource) Name() string {
+	return r.name
+}
+
+// TODO uncomment when Resource interface is redefined.
+//func (r *CRUDResource) Underlying() Resource {
+//	return r
+//}
```

Comment:
I realized we don't need Name() and Underlying() in Ops.

Reply:
I think we do. We have the metrics resource and the retry resource which need to execute the underlying resource and need to know its name. This is how the resources work currently in our operators. Why do you think we do not need this now?

Reply:
We need that. You are right.